Add documentation for CA1872 (#40180)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca1872.md

# CA1872: Prefer 'Convert.ToHexString' and 'Convert.ToHexStringLower' over call chains based on 'BitConverter.ToString'

|                                     | Value                                                                                                         |
| ----------------------------------- |---------------------------------------------------------------------------------------------------------------|
| **Rule ID**                         | CA1872                                                                                                        |
| **Title**                           | Prefer 'Convert.ToHexString' and 'Convert.ToHexStringLower' over call chains based on 'BitConverter.ToString' |
| **Category**                        | [Performance](performance-warnings.md)                                                                        |
| **Fix is breaking or non-breaking** | Non-breaking                                                                                                  |
| **Enabled by default in .NET 9**    | As suggestion                                                                                                 |

## Cause

> A call to <xref:System.BitConverter.ToString*?displayProperty=nameWithType> followed by a call to <xref:System.String.Replace*?displayProperty=nameWithType> to remove dashes is used to encode bytes to a hexadecimal string representation.

Reviewer: ```suggestion
A call to <xref:System.BitConverter.ToString%2A?displayProperty=nameWithType> followed by a call to <xref:System.String.Replace%2A?displayProperty=nameWithType> to remove dashes is used to encode bytes to a hexadecimal string representation.
```

> This rule will also fire if <xref:System.String.ToLower*?displayProperty=nameWithType> is used in the call chain.

Reviewer: ```suggestion
This rule also fires if <xref:System.String.ToLower%2A?displayProperty=nameWithType> is used in the call chain.
```

## Rule description

> Use <xref:System.Convert.ToHexString*?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower*?displayProperty=nameWithType> when encoding bytes to a hexadecimal string representation. These methods are more efficient and allocation-friendly than using <xref:System.BitConverter.ToString*?displayProperty=nameWithType> in combination with  <xref:System.String.Replace*?displayProperty=nameWithType> to remove dashes and <xref:System.String.ToLower*?displayProperty=nameWithType>.

Reviewer: ```suggestion
Use <xref:System.Convert.ToHexString%2A?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower%2A?displayProperty=nameWithType> when encoding bytes to a hexadecimal string representation. These methods are more efficient and allocation-friendly than using <xref:System.BitConverter.ToString%2A?displayProperty=nameWithType> in combination with <xref:System.String.Replace%2A?displayProperty=nameWithType> to remove dashes and <xref:System.String.ToLower%2A?displayProperty=nameWithType>.
```

## How to fix violations

> To fix a violation of this rule, replace the call chain with either <xref:System.Convert.ToHexString*?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower*?displayProperty=nameWithType>.

Reviewer: ```suggestion
To fix a violation of this rule, replace the call chain with either <xref:System.Convert.ToHexString%2A?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower%2A?displayProperty=nameWithType>.
```

    }
}
```

```vb
Imports System
Imports System.Text

Class HelloWorldEncoder
    Private ReadOnly _data As Byte() = Encoding.ASCII.GetBytes("Hello World")

    Public Function Encode() As String
        Return BitConverter.ToString(_data).Replace("-", "")
    End Function

    Public Function EncodeToLower() As String
        Return BitConverter.ToString(_data).Replace("-", "").ToLower()
    End Function
End Class
```

The following code snippet fixes the violation:

```csharp
using System;
using System.Text;

class HelloWorldEncoder
{
    private readonly byte[] _data = Encoding.ASCII.GetBytes("Hello World");

    public string Encode()
    {
        return Convert.ToHexString(data);
    }

    public string EncodeToLower()
    {
        return Convert.ToHexStringLower(data);
    }
}
```

```vb
Imports System
Imports System.Text

Class HelloWorldEncoder
    Private ReadOnly _data As Byte() = Encoding.ASCII.GetBytes("Hello World")

    Public Function Encode() As String
        Return Convert.ToHexString(data)
    End Function

    Public Function EncodeToLower() As String
        Return Convert.ToHexStringLower(data)
    End Function
End Class
```

## When to suppress warnings

> It is safe to suppress a warning from this rule; however, we recommend that you use either <xref:System.Convert.ToHexString*?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower*?displayProperty=nameWithType>.

Reviewer: ```suggestion
It is safe to suppress a warning from this rule; however, we recommend that you use either <xref:System.Convert.ToHexString%2A?displayProperty=nameWithType> or <xref:System.Convert.ToHexStringLower%2A?displayProperty=nameWithType>.
```
